DIPs/DIP1009.md

```diff
@@ -94,34 +94,77 @@ class C {
 
-The drawback of this conciseness is that a usability gap may open up between those who find the built-in `assert` functionality adequate to their needs, and those who, for whatever reason, are forced to use some other system of checking. The existing `in` and `out` contract syntax is still available for use by the latter. But the language now more strongly promotes using existing `assert` functionality where possible. The incentive is increased, therefore, to make `assert` serve as many people as possible, however that is achieved.
+The drawback of this conciseness is that a usability gap may open up between those who find the built-in `assert` functionality adequate to their needs, and those who, for whatever reason, are forced to use some other system of checking. The existing `in` and `out` contract syntax is still available for use by the latter. But the language now more strongly promotes using existing `assert` functionality where possible. The incentive is increased, therefore, to make `assert` serve as many people as possible.
 
 ### Analysis of the new `out` syntax
 
-The most desirable syntax for the new `out` expression, i.e. a single parenthesized set of `assert` arguments,  is ambiguous with the existing `out` contract grammar. The presence of nothing but a single identifier inside the parentheses could be interpreted as either syntax, and thus can't be recommended. All of the proposed alternatives were imperfect in one way or another. One such alternative is to require two sets of parentheses, which resembles template function declarations and the existing `out` contract syntax, e.g.:
+The most desirable syntax for the new `out` expression, i.e. a single parenthesized set of `assert` arguments,  is ambiguous with the existing `out` contract grammar. The presence of nothing but a single identifier inside the parentheses could be interpreted as either syntax. None of the proposed solutions is perfect.
+
+1. The simplest solution is for the new grammar to require the condition to contain additional tokens when checking a non-return identifier. So instead of `out(i)`, for example, one must write `out(i != 0)`, and ambiguity is avoided. Any programmer who forgets to write those extra tokens will be met with a parsing error eventually, if not immediately.
+```d
+int fun(ref int i)
+  out(i != 0) // okay
+  out(i) // whoops! ambiguous with existing contract syntax
+{
+   // out contract, not function body
+}
+... // parsing error inevitable
+```
+The compiler can possibly use extra state logic to detect this particular error and assist the programmer with a helpful error message, "`out` contract expressions require more than just a single identifier", when the parsing finally fails. Very few exit contracts will, however, check outgoing variables with nothing but a single identifier. Therefore this will be the rare case.
```

Comment:
Another way to disambiguate is to require `do` here: `out(identifier) do {...}` when `identifier` should be interpreted as a boolean contract expression. The compiler could advise this in the error message mentioned above. I think this is clearer than a single leading semi-colon for readability, and more intuitive when reading too. (That's why I like the `r => r < 0` syntax). I would keep the `do` keyword on the same line as the `out` contract to avoid indentation, and this case would be uncommon anyway.